### DISCUSSION TITLE
Is it possible for a UserData method to return a reference to this (self)?

### DISCUSSION THREAD

**Discussion by drauschenbach** [mlua-rs/mlua#357] [Q&A] [answered]
I'm working on a configuration builder type of UserData, where every method call would ideally return the `this` object reference. But I'm having trouble finding a way to get my methods to return `this` instead of some new copy.

My UserData is something like this:

```rust
pub struct MessageBuilder {
  pub topic: Mutex<String>,
  pub partition: Mutex<Option<i32>>,
}

impl UserData for MessageBuilder {
fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method("partition", |lua, this, partition: i32| {
            let mut locked_partition = this
                .partition
                .lock()
                .map_err(|err| Error::RuntimeError(err.to_string()))?;
            *locked_partition = Some(partition);
            Ok(/* WHAT CAN GO HERE? */)
        });
        methods.add_method("topic", ...
}
```

And I want to use it like this:

```lua
local message = MessageBuilder:new():topic("topic"):partition(23):build()
```

**Comment by khvzak** [comment, marked as answer]
You need to use [`add_function`](https://docs.rs/mlua/latest/mlua/trait.UserDataMethods.html#tymethod.add_function) which will receive userdata as first arg in form of `AnyUserData` to pass back.

```rust
impl mlua::UserData for MessageBuilder {
    fn add_methods<'lua, M: LuaUserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_function("partition", |_, (ud, partition): (AnyUserData, _)| {
            let mut this = ud.borrow_mut::<Self>()?;
            this.partition = Some(partition);
            Ok(ud)
        });
    }
}
```

Also you don't really need to mutex to simply change struct fields, as can always borrow underlying data mutably.

**Reply by drauschenbach** [reply]
Excellent. That allows me to perform a [massive cleanup](https://gitlab.com/megalithic-llc/mlua-image/-/merge_requests/9). Thanks!!
